test/widgets/inbox_test.dart

```diff
@@ -310,6 +327,30 @@ void main() {
           checkAppearsUncollapsed(tester, findSectionContent);
         });
 
+        testWidgets('collapse all-DMs section after scroll', (tester) async {
+          await setupVarious(tester);
+
+           final listFinder = find.byType(Scrollable);
+          // Scroll the [StickyHeaderListView] enough so that
+          // the [_AllDmsSection] shows a sticky header
+          await tester.drag(listFinder, const Offset(0, -50));
+
+          final headerRow = findAllDmsHeaderRow(tester);
+          // Check that the sticky header is present
+          check(headerRow).isNotNull();
```

Comment:
One useful tactic for making the test robust (as requested in my previous comment, a few lines below here) would be to insert clearer checks on the test setup here.

This check has a comment saying "Check that the sticky header is present". But it seems to check just that there is *a* header for the DMs section — not that it's behaving as a sticky header.

Relatedly: why is 50px an appropriate amount to scroll by? If something about the setup changed so that 50px was no longer an appropriate amount (maybe the layout got taller, or shorter, or the data in `setupVarious` was edited, or something else), can we have something in the test that would detect that?

Reply:
Yeah, the `50px` is surely not a robust way of scrolling. A more appropriate way would be to scroll until a specific widget is off-screen. I hope there was an opposite method to `WidgetTester().scrollUntilVisible`, something like `scrollUntilInvisible`, in that case, it would be so much easy and error-free when using it. 

For example, if we wanted to test the `_AllDmsSection`, in the `setupVarious` method, we would add DM messages from two different users, like from `otherUser` and `thirdUser` to `selfUser`.  Now in our test, we would write something like this:

```dart
await setupVarious(tester);

final listFinder = find.byType(Scrollable);

final findSectionContent = find.text(eg.otherUser.fullName);
await scrollUntilInvisible(findSectionContent, 50,
              scrollable: listFinder); // The method I wish was available
              
// Now we're sure that the [_AllDmsSection] has a sticky header, as we scrolled by one message.
              
final headerRow = findAllDmsHeaderRow(tester);
check(headerRow).isNotNull();

// Following, we check if the headerRow is still visible after tapping on it.
```

I even tried to replicate the `scrollUntilInvisible` method from the `WidgetTester().scrollUntilVisible` implementation. Internally it calls the `dragUntilVisible` method, so in fact, I created the `dragUntilInvisible` method like the following:

The `WidgetTester().dragUntilVisible` implementation
```dart
Future<void> dragUntilVisible(
    finders.FinderBase<Element> finder,
    finders.FinderBase<Element> view,
    Offset moveStep, {
      int maxIteration = 50,
      Duration duration = const Duration(milliseconds: 50),
  }) {
    return TestAsyncUtils.guard<void>(() async {
      while (maxIteration > 0 && finder.evaluate().isEmpty) {
        await drag(view, moveStep);
        await pump(duration);
        maxIteration -= 1;
      }
      await Scrollable.ensureVisible(element(finder));
    });
  }
```

My implementation for the `dragUntilInvisible` method
```dart
Future<void> dragUntilInvisible(
    finders.FinderBase<Element> finder,
    finders.FinderBase<Element> view,
    Offset moveStep, {
      int maxIteration = 50,
      Duration duration = const Duration(milliseconds: 50),
  }) {
    return TestAsyncUtils.guard<void>(() async {
      while (maxIteration > 0 && finder.evaluate().isNotEmpty) {
        await drag(view, moveStep);
        await pump(duration);
        maxIteration -= 1;
      }
    });
  }
```

I hoped it would work, but it didn't, and the reason is that the `finder.evaluate.isNotEmpty` is always `true`, which means that even if the widget from the `finder` is off-screen, the `finder` still finds it. I tested this behavior in the `test` for the `stream section` too. 

If I write a simple test like this:

```dart
await setupVarious(tester);

final topicFinder = find.text('stream 2 topic 39');

print(topicFinder.evaluate().length);
```
This will print 1 to the console, even though it should print 0, because a topic with the title `stream 2 topic 39` is surely off-screen in the bottom.

Sorry for the long comment, I hope I was clear.

Reply:
> This will print 1 to the console, even though it should print 0, because a topic with the title `stream 2 topic 39` is surely off-screen in the bottom.

Ah. What's happening here is that the whole section is currently a single item in the list: #389

As a result, because part of the `stream2` section is in view, the whole section gets built and laid out, so all the widgets that belong to it are findable by `find.text(…)`.

You can fix that by using `Finder.hitTestable`:
```dart
  testWidgets('tmp', (tester) async {
    await setupVarious(tester);
    final topicFinder = find.text('stream 2 topic 39').hitTestable();
    check(topicFinder.evaluate()).isEmpty();
  });
```

That test passes, while the version without `hitTestable` doesn't.

Reply:
I see. Thanks for the comment.

What're your thoughts about the idea and implementation of `dragUntilInvisible`?

Reply:
Sure, seems reasonable.

Do make sure that if it runs out of iterations without succeeding, that it throws instead of returning. I believe `dragUntilVisible` is getting that effect from the `element(finder)` call.